src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/RandomNumberGenerator.cs

```diff
@@ -0,0 +1,115 @@
+    /// </summary>
+    internal class RandomNumberGenerator
+    {
+        private readonly long[] _seedArray = new long[56];
+        private int _inext;
+        private int _inextp;
+
+        [ThreadStatic] private static RandomNumberGenerator? t_random;
+
+        public static RandomNumberGenerator Current
+        {
+            get
+            {
+                if (t_random == null)
+                {
+                    t_random = new RandomNumberGenerator();
+                }
+                return t_random;
+            }
+        }
+
+        public RandomNumberGenerator() : this(((long)Guid.NewGuid().GetHashCode() << 32) | (long)Guid.NewGuid().GetHashCode()) { }
+
+        public RandomNumberGenerator(long Seed)
+        {
+            long subtraction = (Seed == long.MinValue) ? long.MaxValue : Math.Abs(Seed);
+            long mj = 1618033988749894848L - subtraction; // magic number based on Phi (golden ratio)
+            _seedArray[55] = mj;
+            long mk = 1;
+
+            int ii = 0;
+            for (int i = 1; i < 55; i++)
+            {
+                // The range [1..55] is special (Knuth) and so we're wasting the 0'th position.
+                if ((ii += 21) >= 55)
+                {
+                    ii -= 55;
+                }
+
+                _seedArray[ii] = mk;
+                mk = mj - mk;
+                if (mk < 0)
+                {
+                    mk += int.MaxValue;
+                }
+
+                mj = _seedArray[ii];
+            }
+
+            for (int k = 1; k < 5; k++)
+            {
+                for (int i = 1; i < 56; i++)
+                {
+                    int n = i + 30;
+                    if (n >= 55)
+                    {
+                        n -= 55;
+                    }
+
+                    _seedArray[i] -= _seedArray[1 + n];
+                    if (_seedArray[i] < 0)
+                    {
+                        _seedArray[i] += int.MaxValue;
+                    }
+                }
+            }
+
+            _inextp = 21;
+        }
+
+        public long Next() => InternalSample();
+
+        private long InternalSample()
+        {
+            int locINext = _inext;
+            if (++locINext >= 56)
+            {
+                locINext = 1;
+            }
+
+            int locINextp = _inextp;
+            if (++locINextp >= 56)
+            {
+                locINextp = 1;
+            }
+
+            long retVal = _seedArray[locINext] - _seedArray[locINextp];
+
+            if (retVal == long.MaxValue)
+            {
+                retVal--;
+            }
+
+            if (retVal < 0)
+            {
+                retVal += long.MaxValue;
+            }
+
+            _seedArray[locINext] = retVal;
+            _inext = locINext;
+            _inextp = locINextp;
+
+            return retVal;
+        }
+    }
+}
```

Comment:
I don't like that we're duplicating all of this code.  https://github.com/dotnet/runtime/issues/26741 is already approved to add NextInt64 to Random... can we just do that and then use it here?  Or just call Next twice?  If for some reason we need to code a custom generator, there are better/faster choices, but it's not clear to me that's needed.

Reply:
The underlying issue that needs to be solved is that the internal state of Random doesn't start with enough entropy in it because it is constructed from a 32 bit seed. https://github.com/dotnet/runtime/pull/46890#discussion_r556235751. If that problem can be resolved without needing a separate copy of the algorithm thats all the better, but I wasn't counting on that being viable given back-compat expectations.

Reply:
A little more info:
1. There is no requirement for reproducibility or consistency of implementation over time here, the numbers just need to be collision resistant.
2. DiagnosticSource.dll ships as an OOB package that can be run on downlevel versions of .NET. We could have different PRNG solutions on different versions if we needed to.

Reply:
> The underlying issue that needs to be solved is that the internal state of Random doesn't start with enough entropy in it because it is constructed from a 32 bit seed

Do we know that just replacing int with long in a copy of the Random code actually meaningfully improves things?  Yes, the seed can now be 64-bit.  Does the rest of the math work out, for a good enough distribution to make your concerns go away?
cc: @tannergooding 

If memory serves, we don't actually like the algorithm used by Random, in that it's both slower and produces lower-quality pseudo-randomness than other algorithms developed in the interim.  We've always just been hesitant to change it due to concerns about compatibility.

Reply:
> Do we know that just replacing int with long in a copy of the Random code actually meaningfully improves things? 

Yes, it does and it has very big difference. From my local testing when using `int`, the period (which the random number can repeat from same generator) is shorter than when using long. When using int this can be less than 100,000. When using long this can happen after 2 billion. 

Also, the algorithm used is called "Additive number generator" described in Knuth book and from the reading which is not tied to specific word size but it is generic and it is expected to get better results if you have bigger word size.

Reply:
> Does the rest of the math work out, for a good enough distribution to make your concerns go away?

As I just commented on https://github.com/dotnet/runtime/issues/26741, my biggest concern with the `long` random APIs is due to using `double` for scaling the sample to the required range.

I've not looked in depth at this implementation, but it isn't using double anywhere, so my biggest concern is gone 😄

Reply:
> If memory serves, we don't actually like the algorithm used by Random, in that it's both slower and produces lower-quality pseudo-randomness than other algorithms developed in the interim

I'm glad for us to use any algorithm that we believe is more recommended. I only investigated our existing Random implementation far enough to note that modifying it appeared like a plausible solution. I hadn't made any effort to survey the new state of the art for PRNGs in general.

Reply:
Just as an example, I translated `Xoshiro256**` from https://en.wikipedia.org/wiki/Xorshift to C#:
```C#
internal class Xoshiro256StarStar
{
	private ulong _s0, _s1, _s2, _s3;

	public unsafe Xoshiro256StarStar()
	{
		while ((_s0 | _s1 | _s2 | _s3) == 0)
		{
			Guid g1 = Guid.NewGuid();
			Guid g2 = Guid.NewGuid();
			ulong* g1p = (ulong*)&g1;
			ulong* g2p = (ulong*)&g2;
			_s0 = *g1p;
			_s1 = *(g1p + 1);
			_s2 = *g2p;
			_s3 = *(g2p + 1);
		}
	}

	private ulong Rol64(ulong x, int k) => (x << k) | (x >> (64 - k));

	public long Next()
	{
		ulong result = Rol64(_s1 * 5, 7) * 9;
		ulong t = _s1 << 17;

		_s2 ^= _s0;
		_s3 ^= _s1;
		_s1 ^= _s2;
		_s0 ^= _s3;

		_s2 ^= t;
		_s3 = Rol64(_s3, 45);

		return (long)result;
	}
}
```
The code is much more concise than the Random code, and it's more than twice as fast:

|         Method |     Mean |     Error |    StdDev |
|--------------- |---------:|----------:|----------:|
| NextXoshiro256 | 3.194 ns | 0.0095 ns | 0.0074 ns |
|   NextRandom64 | 7.644 ns | 0.0916 ns | 0.0857 ns |

while also involving much less allocation for the instance.  It's also reported to have a very good pseudo-random qualities, though I'm admittedly not an expert.

That's all to say, if we're really keen on including a custom generator here, we might as well do it "right".

Reply:
(That said, I'm still keen on having the .NET Core implementation not use its own generator, and instead rely on System.Random, if at all possible.)

Reply:
> (That said, I'm still keen on having the .NET Core implementation not use its own generator, and instead rely on System.Random, if at all possible.)

What you suggest to do when running on NetFx?

Reply:
> Just as an example, I translated Xoshiro256** from en.wikipedia.org/wiki/Xorshift to C#:

If believe it is more reliable, I can give it a try and test it.

Reply:
> What you suggest to do when running on NetFx?

Continue using Guid.NewGuid() as we are today.

Reply:
Also, does `Xoshiro256` always return positive values?

Reply:
> Also, does Xoshiro256 always return positive values?

It covers the full 64-bit space, so if returning long rather than ulong, yes.

But, to be clear, I picked that as an example after about 1 minute of research.  I'm not suggesting it's actually the best option, just highlighting that our System.Random implementation, well, isn't :)

Reply:
> Continue using Guid.NewGuid() as we are today.

What we'll do when running against .NET Core 3.1 on Linux?

Reply:
> What we'll do when running against .NET Core 3.1/netstandard on Linux?

For .NET Core 3.1, continue using Guid.NewGuid() as we are today.

For .NET Standard, don't we type forward to a built-in implementation when on .NET Core?  Or is it only an OOB?

Reply:
Again, I'm not dictating here, I'm suggesting we can do the minimum viable for downlevel.  You define what viable is.

Reply:
> For .NET Standard, don't we type forward to a built-in implementation when on .NET Core? Or is it only an OOB?

Yes it is OOB. netstandard.dll don't include any type forward for the types inside this library. 

> For .NET Core 3.1, continue using Guid.NewGuid() as we are today.

I am seeing some customers still care using this library on Core 3.1/5.0. Although this specific issue is not discussed in their scenarios, But the perf is important in general for them.

> Again, I'm not dictating here, I'm suggesting we can do the minimum viable for downlevel. You define what viable is.

Understood. your feedback is very valuable. That is why I was asking more questions to get what I could be missing. 

I want to say I picked the same algorithm used in Random because it has been used for many years and the private implementation I did showing good performance for the scenario we are trying to fix. I already looked at other algorithms which claim it is better (in perf, size, detection...etc.) e.g. https://www.pcg-random.org/ but I hesitated to go this route as I am not sure if the claims are true and if there is any issue with licensing. I don't mind to give such algorithms a try though.

I am inclining to stick with the idea of the custom PRNG regardless which algorithm we use as long as it give us what we need. The reason is, trying to get different behavior on different targeting will make it more complicated for not much benefit. In the future if we start dropping some old targets, we can try to use Random at that time. I am still welcoming any more feedback.

@noahfalk what you think?

Reply:
Just to tell, I gave `Xoshiro256StarStar` a try and seams give a good results as customized Random one.

Reply:
Given what I've now read online on Xoshiro256** and the positive empirical results I'd be glad to have us use it. Licensing seemed easy as [the code](http://prng.di.unimi.it/xoshiro256starstar.c) is public domain. 

In terms of using a custom PRNG vs. using an in-box one, I agree with your line of thinking Tarek. If we have any scenario where we have to rely on the custom PRNG and we believe the custom one is fast/high quality then it is unclear to me what value is gained by having any alternate implementation option such as using the built-in Random. Potentially it saves some code size but given how compact Xoshiro is that is probably a negligible difference.

There is some (I think fairly small) risk that in the future we'd discover a deficiency in the Xoshiro PRNG algorithm but as long as we avoid making the API public or introducing a new requirement for reproducibility in the IDs we'd have lots of flexibility to change implementation in the future if needed.

Reply:
Thanks all for the feedback. I have updated the implementation and pated the final perf numbers.

Reply:
> it is unclear to me what value is gained by having any alternate implementation option such as using the built-in Random

The value is ensuring that we're improving the core platform and not just implementing one-offs here and there because the core implementation isn't currently as good as we'd like it to be.  In general I want us focused on improving the shared code that everyone uses.

Reply:
I think having this one-off here is ok for the sake of benefit when running on Linux with .NET down level versions. In the future we can use the Random when we drop other targets. I am not seeing any conflict with that.

Reply:
@stephentoub would it make sense in the PR https://github.com/dotnet/runtime/pull/47085 to move the `Xoshiro256StarStarImpl` to a common folder and we can include it here? doing that will ensure we still using the shared one and not forking. what you think?

Reply:
> I think having this one-off here is ok for the sake of benefit when running on Linux with .NET down level versions. In the future we can use the Random when we drop other targets. I am not seeing any conflict with that.

The conflict is when the effort is only put into the one-off and not into fixing the core thing.  If everyone does that, the core thing is never fixed.  And seeing that happen is why I went and did it myself in #47085.

> move the Xoshiro256StarStarImpl and ImplBase to a common folder and we can include it here?

Sure

Reply:
Thanks @stephentoub, if you are going to move `Xoshiro256StarStarImpl` to the common source folder, could you please add the following

```C#
#if ALLOW_PARTIALLY_TRUSTED_CALLERS
        [System.Security.SecuritySafeCriticalAttribute]
#endif
```

on the constructor as it is unsafe and that security attribute is important to have when running on the Ful Framework. 

Please let me know if you prefer merging your PR first and then I move and edit the file myself. Anyway, I have to wait you merge your PR first.

Reply:
Thanks.  Go ahead with your change.  Once mine is merged in, you can then tweak it as needed to work well for netfx and the other assembly compilation, and dedup.

Reply:
great. could anyone of you approve the current change if there is no any other comment?